OSNetworkRequirementChecker-HLApp/src/dns-helper.c

```diff
@@ -0,0 +1,523 @@
+    si.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    ret = sendto(fd, queryBuf, (size_t)messageSize, 0, (struct sockaddr *)&si, sizeof(si));
+    if (ret == -1) {
+        Log_Debug("ERROR: sendto: %s (%d)\n", strerror(errno), errno);
+        return -1;
+    }
+    return 0;
+}
+
+int SendServiceDiscoveryQuery(const char *dName, int fd)
+{
+    return SendDnsQuery(dName, ns_c_in, ns_t_ptr, fd);
+}
+
+int SendServiceInstanceDetailsQuery(const char *instanceName, int fd)
+{
+    return SendDnsQuery(instanceName, ns_c_in, ns_t_any, fd);
+}
+
+int SendARecordQuery(const char *domainName, int fd)
+{
+    return SendDnsQuery(domainName, ns_c_in, ns_t_a, fd);
+}
+
+int ProcessMessageBySection(char *buf, ssize_t len, ns_msg msg, ns_sect section,
+                            ServiceInstanceDetails **instanceDetails)
+{
+    char displayBuf[DISPLAY_BUF_SIZE];
+    ns_rr rr;
+    int messageCount = ns_msg_count(msg, section);
+
+    if (!(*instanceDetails)) {
+        // Allocate for ServiceInstanceDetails and initialize its members.
+        *instanceDetails = malloc(sizeof(ServiceInstanceDetails));
+        if (!(*instanceDetails)) {
+            Log_Debug("ERROR: recvfrom: %s (%d)\n", strerror(errno), errno);
+            return -1;
+        }
+        (*instanceDetails)->name = NULL;
+        (*instanceDetails)->host = NULL;
+        (*instanceDetails)->ipv4Address.s_addr = INADDR_NONE;
+        (*instanceDetails)->port = 0;
+        (*instanceDetails)->txtData = NULL;
+        (*instanceDetails)->txtDataLength = 0;
+        (*instanceDetails)->alias = NULL;
+    }
+
+    // Parse each message
+    for (int i = 0; i < messageCount; ++i) {
+        if (ns_parserr(&msg, section, i, &rr)) {
+            Log_Debug("ERROR: ns_parserr: %s (%d)\n", strerror(errno), errno);
+            return -1;
+        }
+        switch (ns_rr_type(rr)) {
+        case ns_t_ptr: {
+            int compressedNameLength =
+                dn_expand(buf, buf + len, ns_rr_rdata(rr), displayBuf, sizeof(displayBuf));
+            if (compressedNameLength > 0 && !(*instanceDetails)->name) {
+                (*instanceDetails)->name = strdup(displayBuf);
+                if (!(*instanceDetails)->name) {
+                    Log_Debug("ERROR: strdup for instance name failed: %s (%d)\n", strerror(errno),
+                              errno);
+                    return -1;
+                }
+            }
+            break;
+        }
+        case ns_t_srv: {
+            // Parse the SRV record and populate the port and host fields in instance details as per
+            // DNS SRV record specification: https://tools.ietf.org/rfc/rfc2782.txt
+            // SRV record format: Priority|  Weight |   Port  |     Target
+            //                   (2 Bytes)|(2 Bytes)|(2 Bytes)|(Remaining Bytes)
+            const char *data = ns_rr_rdata(rr);
+            (*instanceDetails)->port =
+                (uint16_t)ns_get16(data + sizeof(uint16_t) + sizeof(uint16_t));
```

Comment:
This needs removing - it should only be set if the port/host are unset.

Reply:
oh yes I missed that. Fixed.